src/main/java/seedu/address/model/book/BookStatus.java

```diff
@@ -0,0 +1,172 @@
+import static seedu.address.model.book.BookStatusType.BORROWED;
+
+import java.text.ParseException;
+import java.text.SimpleDateFormat;
+import java.util.Date;
+import java.util.Optional;
+
+import seedu.address.model.patron.Patron;
+
+/**
+ * Represents a BookStatus in LibTask.
+ * Guarantees: immutable; is valid as declared in {@link #isValidStatus(BookStatusType, Optional, Optional, Optional)}
+ */
+public class BookStatus {
+
+    public static final String MESSAGE_CONSTRAINTS =
+            "An available book status must not contain any borrower, borrowDate and return date "
+                    + "and a borrowed status must contain borrower, borrowDate and return date,"
+                    + "and date must be in dd-MMM-yyyy format.";
+    public static final SimpleDateFormat STATUS_DATE_FORMAT = new SimpleDateFormat("dd-MMM-yyyy");
+    public static final String VALIDATION_REGEX = "^\\d{2}-[a-zA-Z]{3}-\\d{4}$";
+
+    private final BookStatusType bookStatusType;
+    private final Optional<Patron> borrower;
+    private final Optional<Date> borrowDate;
+    private final Optional<Date> returnDate;
+
+    /**
+     * Constructs a {@code BookStatus}.
+     * @param bookStatusType The status of this book.
+     * @param borrower The patron who borrows this book, if any.
+     * @param borrowDate The borrowDate of this book, if any.
+     * @param returnDate The returnDate of this book, if any.
+     */
+    public BookStatus(BookStatusType bookStatusType, Optional<Patron> borrower, Optional<String> borrowDate,
+                      Optional<String> returnDate) {
+        requireAllNonNull(bookStatusType, borrower, borrowDate, returnDate);
+        checkArgument(isValidStatus(bookStatusType, borrower, borrowDate, returnDate), MESSAGE_CONSTRAINTS);
+        this.borrowDate = borrowDate.map(date -> parseToDate(date));
+        this.returnDate = returnDate.map(date -> parseToDate(date));
+        this.bookStatusType = bookStatusType;
+        this.borrower = borrower;
+    }
+
+    /**
+     * A static function for creating an AVAILABLE book status.
+     *
+     * @return An AVAILABLE BookStatus object.
+     */
+    public static BookStatus createAvailableBookStatus() {
+        return new BookStatus(AVAILABLE, Optional.empty(), Optional.empty(),
+                Optional.empty());
+    }
+
+    /**
+     * Returns true if a given parameters can be converted to a valid book status.
+     */
+    public static boolean isValidStatus(BookStatusType bookStatusType, Optional<Patron> borrower,
+                                        Optional<String> borrowDate, Optional<String> returnDate) {
+        requireAllNonNull(bookStatusType, borrower, borrowDate, returnDate);
+
+        // A book with available status should not have any borrower, borrowing or returning dates
+        boolean isAvailableConstraintsSatisfied = bookStatusType == AVAILABLE
+                && borrower.isEmpty() && borrowDate.isEmpty() && returnDate.isEmpty();
```

Comment:
Elegant code to check that a book with available status should not have any borrower, borrowing or returning dates. The code is easy to understand by just reading from left to right :100: